src/isolation/tempdir.rs

```diff
@@ -1,25 +1,114 @@
 ///
+/// To keep the temporary directory clean and because we need to include the
+/// parent directory in [`crate::isolation::landlock::initialize`], a sub-folder
+/// called `uhyve-{uid}` (e.g. `/tmp/uhyve-1000`) is used to create temporary
+/// directories under.
+///
+/// After creating a directory, the permissions are set to 0o700 (read, write, execute
+/// for the user). Those permissions are checked during the runtime.
+///
 /// * `dir_path` - The location in which the temporary directory should be created.
-pub fn create_temp_dir(dir_path: &Option<String>) -> TempDir {
-	let dir = Builder::new()
+pub fn create_temp_dir(dir_path: Option<&PathBuf>) -> TempDir {
+	Builder::new()
 		.permissions(Permissions::from_mode(0o700))
 		.prefix("uhyve-")
 		.suffix(&Uuid::new_v4().to_string())
-		.tempdir_in(
-			dir_path
-				.as_ref()
-				.map(PathBuf::from)
-				.unwrap_or_else(env::temp_dir),
-		)
-		.expect("The temporary directory could not be created.");
+		.tempdir_in(dir_path.cloned().unwrap_or_else(|| {
+			let mut env_temp_dir = env::temp_dir();
+			// Sane environments presumed to have world-writable temporary directory.
+			debug_assert!(!env_temp_dir.metadata().unwrap().permissions().readonly());
+
+			// getpid should never fail or modify errno, unsafe call is fine.
+			env_temp_dir.push(format!("uhyve-{}", unsafe { libc::getuid() }));
+
+			// If the subfolder doesn't exist, try to create it.
+			// If that isn't possible, panic.
+			match env_temp_dir.metadata() {
+				Ok(metadata) => {
+					// Ensure permissions of directory are correct.
+					// Implicitly also checks if this is a directory.
+					let temp_dir_mode = metadata.permissions().mode();
+					assert_eq!(
+						temp_dir_mode, 0o40700,
+						"Uhyve's preexisting tempdir has incorrect mode ({temp_dir_mode:#o}, expected: 0o40700)."
+					);
+				}
+				Err(ref e) if e.kind() == std::io::ErrorKind::NotFound => {
+					DirBuilder::new()
+						.mode(0o700)
+						.create(&env_temp_dir)
+						.unwrap_or_else(|e| {
+							panic!(
+								"Could not create temporary directory {:?}: {e}",
+								env_temp_dir.as_path(),
+							)
+						});
+				}
+				Err(e) => panic!(
+					"Could not create temporary directory {:?}: {e}",
+					env_temp_dir.as_path()
+				),
+			}
+
+			env_temp_dir
+		}))
+		.expect("The temporary directory could not be created.")
+}
+
+#[cfg(test)]
+mod tests {
+	use std::fs::remove_dir;
+
+	use super::*;
+
+	#[test]
+	fn test_create_with_dir_path() {
+		let env_tempdir = env::temp_dir();
+		let tempdir = create_temp_dir(Some(&env_tempdir));
+		// Assertions deferred for later to clean up directory first.
+		// e.g. /tmp/uhyve-1000/uhyve-8Mubdtc4246409-1913-4123-b4b4-88cb4953a1ea
+		let in_tmp_path: bool = tempdir.path().starts_with(&env_tempdir);
+		let in_tmp_subfolder: bool = tempdir
+			.path()
+			.starts_with(env_tempdir.join(format!("uhyve-{}", unsafe { libc::getuid() })));
+
+		remove_dir(tempdir.path()).unwrap_or_else(|e| {
```

Comment:
Maybe you could factor out "creating a temporary directory using `create_temp_dir`, removing that directory, and only storing/returning the path of it for inspection / assertion" into a utility function inside this `mod tests` namespace, but outside of a specific test.

Reply:
I see your point (that's something I've done for the integration tests), but given that there are only two tests I am not sure if that's helpful. Additionally, the panic message differs between the two instances; I could pass the closure inside of such a helper, but that may be a bit much for a test that won't make the release binary.